### ISSUE TITLE
False-positive "ignoring deprecated linker" warning

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#158192]
When using Zig's CC wrapper as the linker, warn(linker_messages) shows:

```
warning: linker stderr: ignoring deprecated linker optimization setting '1'
  |
  = note: `#[warn(linker_messages)]` on by default
```

This is coming from here: https://codeberg.org/ziglang/zig/src/commit/e949341b705c16b46544f26f5aab0a16670f42b4/src/main.zig#L3245

I suspect we should either silence this message unconditionally in our existing filtering code or figure out a detection strategy for users opting into using Zig (and then not pass -O flags).

cc https://github.com/rust-lang/rust/issues/136096

**Comment by jyn514** [comment]
i think this is working as intended? where did you run into this issue?

**Comment by Mark-Simulacrum** [comment]
> i think this is working as intended?

I think it's working as expected based on the current implementation, but I don't think our intent is that users using "reasonably common" linker configurations see warnings. In this case rustc is the one passing -O (https://github.com/rust-lang/rust/blob/f28ac764c36004fa6a6e098d15b4016a838c13c6/compiler/rustc_codegen_ssa/src/back/linker.rs#L711), so rustc is responsible for this warning.

> where did you run into this issue?

As I mentioned in the description, this is coming from a Zig toolchain being used for linking (Cargo `linker` configured as equivalent of `zig cc  "$@" --target=x86_64-linux-musl`, though I think the exact extra bits don't matter).

We could call this a bug in Zig if we think it shouldn't emit warnings on `-O` being deprecated for zig's linker.

**Comment by mati865** [comment]
Warning about `-O` arg is questionable IMO. All build systems add that because otherwise most of the linkers default to `-O0` which rarely is what you want.
So, I suppose they are fine with this warning being printed every time using zig-cc with a build system.
